app/src/main/java/com/tari/android/wallet/service/WalletService.kt

```diff
@@ -592,8 +601,52 @@ class WalletService : Service(), FFIWalletListenerAdapter {
             return pendingOutboundTx
         }
 
-        // endregion
+        override fun requestTestnetTari() {
+            val message = "$MESSAGE_PREFIX $publicKeyHexString"
+            val signature = wallet.signMessage(message)
+            val requestBody = TestnetTariAllocateRequest(signature, publicKeyHexString)
 
-    }
+            val response = tariService.getTransaction(publicKeyHexString, requestBody)
+            response.enqueue(object : Callback<TestnetTariAllocateResponse> {
+                override fun onFailure(call: Call<TestnetTariAllocateResponse>, t: Throwable) {
+                    notifyFreeTariAllocationFailed(getString(R.string.service_error_no_internet_connection))
+                }
 
+                override fun onResponse(
+                    call: Call<TestnetTariAllocateResponse>,
+                    response: Response<TestnetTariAllocateResponse>
+                ) {
+                    when (response.code()) {
+                        in 200..209 -> {
+                            response.body()?.let {
+                                val publicKey = FFIPublicKey(HexString(it.return_wallet_id))
+                                val privateKey = FFIPrivateKey(HexString(it.key))
+                                val value = BigInteger(it.value)
+                                val ffiContact = FFIContact("TariBot", publicKey)
+
+                                wallet.addUpdateContact(ffiContact)
+                                wallet.importUTXO(
+                                    value,
+                                    getString(R.string.home_tari_bot_some_tari_to_get_started),
+                                    privateKey,
+                                    publicKey
+                                )
+                                listeners.iterator().forEach {
+                                    it.onTestnetTariRequestSuccess(true)
+                                }
+                            }
+                        }
+                        else -> notifyFreeTariAllocationFailed(getString(R.string.service_error_wallet_could_not_allocate_free_tari))
+                    }
+                }
+            })
+        }
+
+        private fun notifyFreeTariAllocationFailed(error: String) {
```

Comment:
Free -> Testnet. You can also replace "allocate" with request as talked earlier.